src/cmdline/mod.rs

```diff
@@ -209,6 +220,114 @@ impl Cmdline {
+    pub fn add_console<T: AsRef<str>>(&mut self, device: T, options: Vec<T>) -> Result<()> {
+        if device.as_ref().is_empty() {
+            return Err(Error::InvalidDevice(String::new()));
+        }
+        let mut serial_str = "console=".to_string();
+        serial_str.push_str(device.as_ref());
+        for opt in options {
+            serial_str.push_str(",");
+            serial_str.push_str(opt.as_ref());
+        }
+
+        self.insert_str(&serial_str)
+    }
+
+    /// Adds a virtio MMIO device to the kernel command line.
+    ///
+    /// Multiple devices can be specified, with multiple `virtio_mmio.device=` options. This
+    /// function must be called once per device.
+    /// The function appends a string of the following format to the kernel command line:
+    /// `<size>@<baseaddr>:<irq>[:<id>]`.
+    /// For more details see the [documentation] (section `virtio_mmio.device=`).
+    ///
+    /// # Arguments
+    ///
+    /// * `size` - size of the slot the device occupies on the MMIO bus.
+    /// * `baseaddr` - physical base address of the device.
+    /// * `irq` - interrupt number to be used by the device.
+    /// * `id` - optional platform device ID.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use linux_loader::cmdline::*;
+    /// # use std::ffi::CString;
+    /// # use vm_memory::{GuestAddress, GuestUsize};
+    /// let mut cl = Cmdline::new(100);
+    /// cl.add_virtio_mmio_device(1 << 12, GuestAddress(0x1000), 5, Some(42)).unwrap();
+    /// let cl_cstring = CString::new(cl).unwrap();
+    /// assert_eq!(cl_cstring.to_str().unwrap(), "virtio_mmio.device=4K@0x1000:5:42");
+    /// ```
+    ///
+    /// [documentation]: https://www.kernel.org/doc/html/latest/admin-guide/kernel-parameters.html
+    pub fn add_virtio_mmio_device(
+        &mut self,
+        size: GuestUsize,
+        baseaddr: GuestAddress,
+        irq: u32,
+        id: Option<u32>,
+    ) -> Result<()> {
+        if size == 0 {
+            return Err(Error::MmioSize);
+        }
+
+        let mut device_str = format!(
+            "virtio_mmio.device={}@0x{:x?}:{}",
+            Self::guestusize_to_str(size),
+            baseaddr.raw_value(),
+            irq
+        );
+        if let Some(id) = id {
+            device_str.push_str(format!(":{}", id).as_str());
+        }
+        self.insert_str(&device_str)
+    }
+
+    // Converts a `GuestUsize` to a concise string representation, with multiplier suffixes.
+    fn guestusize_to_str(size: GuestUsize) -> String {
+        const KB_MULT: u64 = 1 << 10;
+        const MB_MULT: u64 = KB_MULT << 10;
+        const GB_MULT: u64 = MB_MULT << 10;
+
+        if size % GB_MULT == 0 {
+            return format!("{}G", size / GB_MULT);
+        }
+        if size % MB_MULT == 0 {
+            return format!("{}M", size / MB_MULT);
+        }
+        if size % KB_MULT == 0 {
+            return format!("{}K", size / KB_MULT);
+        }
+        size.to_string()
```

Comment:
Is the size valid if none of the standard suffixes are appropriate? I was thinking if we should return an error in case none of them matches.

Reply:
It's valid. The kernel trims out the suffix, if present, and parses the number normally, so it's valid even if expressed in bytes. No further validations appear to be made.
Parsing the device: https://elixir.bootlin.com/linux/latest/source/drivers/virtio/virtio_mmio.c#L636
Parser: https://elixir.bootlin.com/linux/latest/source/lib/cmdline.c#L125